WolvenKit.Modkit/RED4/Tools/MeshTools.cs

```diff
@@ -661,6 +663,43 @@ public static List<RawMeshContainer> ContainRawMesh(MemoryStream gfs, MeshesInfo
             RemoveDoubleFaces(ref expMeshes);
             return expMeshes;
         }
+
+        private static string FindNextInSequence(IReadOnlyList<string> sequence, string current){
+            for (var i = 0; i < sequence.Count - 1; i++)
+            {
+                if (sequence[i] != current || sequence[i + 1] == current)
+                {
+                    continue;
+                }
+                return sequence[i + 1];
+            }
+            return null;
+        }
+
+        private static string[] InitializeDefaultMaterial(int numTotalSubmeshes, string[] meshAppearances)
+        {
+            List<string> ret = meshAppearances.ToList();;
+            // If the array is empty: initialize it with "default" material
+            if (ret.Count == 0)
+            {
+                ret.Add("default");
+            }
+
+            // append all missing appearance names
+            for (var e = ret.Count; e < numTotalSubmeshes; e++)
+            {
+                // get last item from array, which is never falsy because of the above statement 
+                var lastUsedString = ret[^1];
+
+                var nextUsedString = FindNextInSequence(ret, lastUsedString) ?? lastUsedString;
```

Comment:
Does it need to reuse appearance names to fill in the array?

Reply:
It could also randomly assign something (e.g. default), but I've given an example of how those arrays usually look in the linked [issue](https://github.com/WolvenKit/WolvenKit/issues/1035), and this takes care of all of these cases.

Reply:
Then I would say to just use `e` with condition on delta (i.e. `numTotalSubmeshes - ret.Count`) instead of finding the next in sequence, it should be the same behavior.

Reply:
I'm not familiar enough with c# for that, can you sketch it out for me?

Reply:
Wait, I think that does something different?

Reply:
The way I understand it now is that you do the following:

```
numTotalSubmeshes: 7
meshAppearances: [ mesh1, mesh2, mesh3, mesh4, , mesh5 ]

FindNextInSequence(numTotalSubmeshes, meshAppearances): [ mesh1, mesh2, mesh3, mesh4, , mesh5, mesh1, mesh2 ]
```

This way I do not see a reason why there should be 2 for-loops, your entire `InitializeDefaultMaterial` is having a worst-case complexity of `O(n^2)`, assume this:

```
numTotalSubmeshes: 7
meshAppearances: [ mesh1 ]
```

When you call `InitializeDefaultMaterial` your algorithm will start doing more comparisons with each element you add to `ret`, first it will do 1, then 2, then 3, ... then 6.

Another example of output is:

```
numTotalSubmeshes: 7
meshAppearances: [ mesh1, mesh2, mesh3 ]

FindNextInSequence(numTotalSubmeshes, meshAppearances): [ mesh1, mesh2, mesh3, mesh1,  mesh2, mesh1, mesh2 ]
```

Note that `mesh3` is not used anymore, which I assume is not expected behavior.

So what I was proposing was something like this:

```
for (int i = 0; i < (numTotalSubmeshes - ret.Count); i++)
{
    var newItem = ret[i];
    ret.Add(newItem);
}
```
I should do the same thing you are doing now.

Or am I not fully understanding what you want to do here?

Reply:
Ops, my bad the second statement (about `mesh3`) is not valid.

Reply:
I get what you mean now, but that would always start at the beginning. Means if I'd extend a list to 5 entries, your solution would lead to
```
[ mat1, mat2, mat3, mat1 ] => [ mat1, mat2, mat3, mat1, mat1 ]
```
and mine should be
```
[ mat1, mat2, mat3, mat1 ] => [ mat1, mat2, mat3, mat1, mat2 ]
```
... since it checks the last used item in the sequence.

I'm sure you could achieve the same sorcery with lambdas and streams somehow, but I don't know how ':)

Reply:
Ops, my bad the second statement (about `mesh3`) is not correct, should be another output, that is `[ mesh1, mesh2, mesh3, mesh3,  mesh3, mesh3, mesh3 ]`.

Reply:
Yeah, my idea was since the original meshes are set up to always repeat the same sequence, the autocomplete should for reasons of maximum convenience do the same.
It might probably be better to move all this to the mesh import section so it can get written to disk? But I couldn't find that. :D Maybe I'll do that next.

Reply:
> I get what you mean now, but that would always start at the beginning. Means if I'd extend a list to 5 entries, your solution would lead to
> 
> ```
> [ mat1, mat2, mat3, mat1 ] => [ mat1, mat2, mat3, mat1, mat1 ]
> ```
> 
> and mine should be
> 
> ```
> [ mat1, mat2, mat3, mat1 ] => [ mat1, mat2, mat3, mat1, mat2 ]
> ```
> 
> ... since it checks the last used item in the sequence.
> 
> I'm sure you could achieve the same sorcery with lambdas and streams somehow, but I don't know how ':)

Ah, I see now. But then I would a for-loop before appending to the list to get the last index of the mesh, e.g.:

```csharp
var lastIndex = ret.TakeWhile(s => s == ret[^1]).Count();

// TODO: Do all the checks for out of bounds.

for (var e = ret.Count; e < numTotalSubmeshes; e++)
{
    ret.Add(ret[lastIndex++]);
}
```

Reply:
That won't update, though, will it? Like... it might be overkill to recalculate the next sequence item each time. Maybe we could figure out the sequence length, and use that as an offset? but it'd require lots of out-of-bounds checking, and might not improve read- and maintainability... I'm uncertain.

Reply:
Not sure what you mean with `won't update`, but it won't be an overkill, you can find the index of the last entry with `IndexOf`, then assume everything before it is already correct.

There is two cases that comes to my mind now: 
1. `startIndex + 1 == list.Count -> startIndex = 0`, if this happen you can assume this is the last entry in the list and can safely start from 0.
3. An input like `[ mesh1, mesh2, mesh3, mesh3 ]` will cause issues with `IndexOf`. But maybe just assume this is wrong / intended and just append `mesh3` (penultimate value).

Anyway, my point here is that the speed of the algorithm can be improved.

Reply:
You're absolutely right. I have an idea how I can get that out of the loop. I'll sleep on it and hopefully implement it tomorrow :)

Reply:
Please do not sleep on it, it is fragile.

Reply:
If it was C...

Reply:
Pushed update, works. Please resolve and proceed at your discretion :))